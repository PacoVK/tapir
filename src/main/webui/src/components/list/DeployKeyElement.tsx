import { DeployKey } from "../../types";
import {
  Avatar,
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogContentText,
  DialogTitle,
  IconButton,
  ListItem,
  ListItemAvatar,
  ListItemText,
  Tooltip,
  Typography,
} from "@mui/material";
import DeleteIcon from "@mui/icons-material/Delete";
import CopyIcon from "@mui/icons-material/ContentCopy";
import LoopIcon from "@mui/icons-material/Loop";
import KeyIcon from "@mui/icons-material/Key";
import React, { useState } from "react";
import { formatDateTime } from "../../util/DateUtil";

type Action = "delete" | "regenerate";

const DeployKeyElement = ({
  deployKey,
  onCopy,
  onRegenerate,
  onDelete,
}: {
  deployKey: DeployKey;
  onCopy: (deployKey: string) => void;
  onRegenerate: (deployKey: string) => void;
  onDelete: (deployKey: string) => void;
}) => {
  const [open, setOpen] = useState(false);
  const [action, setAction] = useState(undefined as Action | undefined);
  const [dialogTitle, setDialogTitle] = useState("");
  const [dialogMessage, setDialogMessage] = useState("");

  const handleDialogClose = () => {
    setAction(undefined);
    setOpen(false);
  };

  const openDeleteKeyDialog = () => {
    setDialogTitle(`Delete Deploykey ${deployKey.id}?`);
    setDialogMessage(
      `Are you sure you want to delete ${deployKey.id}? This is a destructive action and cannot be undone.`,
    );
    setAction("delete");
    setOpen(true);
  };

  const openRegenerateKeyDialog = () => {
    setDialogTitle(`Regenerate Deploykey for ${deployKey.id}?`);
    setDialogMessage(
      `Are you sure you want to regenerate ${deployKey.id}? You'll need to update the key in your CI/CD pipeline.`,
    );
    setAction("regenerate");
    setOpen(true);
  };

  const handleDestructiveAction = () => {
    switch (action) {
      case "regenerate":
        onRegenerate(deployKey.id);
        setOpen(false);
        break;
      case "delete":
        onDelete(deployKey.id);
        setOpen(false);
        break;
      default:
        setOpen(false);
        break;
    }
  };

  const copyToClipboard = async () => {
    await navigator.clipboard.writeText(deployKey.key);
    onCopy(deployKey.id);
  };

  return (
    <>
      <ListItem>
        <ListItemAvatar>
          <Avatar>
            <KeyIcon />
          </Avatar>
        </ListItemAvatar>
        <ListItemText
          primary={deployKey.id}
          secondary={
            <>
              <Typography component={"span"} variant={"subtitle1"}>
                {deployKey.key}
              </Typography>
              <br />
              <Typography component={"span"} variant={"subtitle2"}>
                Last modified at: {formatDateTime(deployKey.lastModifiedAt)}
              </Typography>
            </>
          }
        />
        <Tooltip title="Copy to clipboard">
          <IconButton
            edge="end"
            sx={{ mr: 1 }}
            aria-label="copy"
            onClick={copyToClipboard}
          >
            <CopyIcon />
          </IconButton>
        </Tooltip>
        <Tooltip title="Regenerate">
          <IconButton
            edge="end"
            sx={{ mr: 1 }}
            aria-label="regenerate"
            value={deployKey.id}
            onClick={openRegenerateKeyDialog}
          >
            <LoopIcon />
          </IconButton>
        </Tooltip>
        <Tooltip title="Delete">
          <IconButton
            edge="end"
            aria-label="delete"
            value={deployKey.id}
            onClick={openDeleteKeyDialog}
          >
            <DeleteIcon />
          </IconButton>
        </Tooltip>
      </ListItem>
      <Dialog
        open={open}
        onClose={handleDialogClose}
        aria-labelledby="alert-dialog-title"
        aria-describedby="alert-dialog-description"
      >
        <DialogTitle id="alert-dialog-title">
          <Typography>{dialogTitle}</Typography>
        </DialogTitle>
        <DialogContent>
          <DialogContentText id="alert-dialog-description">
            {dialogMessage}
          </DialogContentText>
        </DialogContent>
        <DialogActions>
          <Button onClick={handleDestructiveAction}>Confirm</Button>
          <Button onClick={handleDialogClose} autoFocus>
            Cancel
          </Button>
        </DialogActions>
      </Dialog>
    </>
  );
};

export default DeployKeyElement;
