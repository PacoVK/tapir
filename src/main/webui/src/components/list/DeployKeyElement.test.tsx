import DeployKeyElement from "./DeployKeyElement";
import { render } from "@testing-library/react";
jest.mock("../../util/DateUtil", () => ({
  formatDateTime: () => "10-10-2021 10:10",
}));

describe("<DeployKeyElement /> spec", () => {
  it("renders the DeployKeyElement", () => {
    const view = render(
      <DeployKeyElement
        deployKey={{
          id: "foo/bar/baz",
          key: "sdlshdfkjlf",
          lastModifiedAt: "2021-10-10T10:10:10.000Z",
        }}
        onCopy={() => {}}
        onDelete={() => {}}
        onRegenerate={() => {}}
      />,
    );
    expect(view).toMatchSnapshot();
  });
});
