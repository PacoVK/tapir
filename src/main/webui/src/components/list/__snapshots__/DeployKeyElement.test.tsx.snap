// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DeployKeyElement /> spec renders the DeployKeyElement 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
      >
        <div
          class="MuiListItemAvatar-root css-1e9lk82-MuiListItemAvatar-root"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-2s90m6-MuiAvatar-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="KeyIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M21 10h-8.35C11.83 7.67 9.61 6 7 6c-3.31 0-6 2.69-6 6s2.69 6 6 6c2.61 0 4.83-1.67 5.65-4H13l2 2 2-2 2 2 4-4.04L21 10zM7 15c-1.65 0-3-1.35-3-3s1.35-3 3-3 3 1.35 3 3-1.35 3-3 3z"
              />
            </svg>
          </div>
        </div>
        <div
          class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            foo/bar/baz
          </span>
          <p
            class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
          >
            <span
              class="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
            >
              sdlshdfkjlf
            </span>
            <br />
            <span
              class="MuiTypography-root MuiTypography-subtitle2 css-ati914-MuiTypography-root"
            >
              Last modified at: 
              10-10-2021 10:10
            </span>
          </p>
        </div>
        <button
          aria-label="copy"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-uof0vn-MuiButtonBase-root-MuiIconButton-root"
          data-mui-internal-clone-element="true"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ContentCopyIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12V1zm3 4H8c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h11c1.1 0 2-.9 2-2V7c0-1.1-.9-2-2-2zm0 16H8V7h11v14z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="regenerate"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-uof0vn-MuiButtonBase-root-MuiIconButton-root"
          data-mui-internal-clone-element="true"
          tabindex="0"
          type="button"
          value="foo/bar/baz"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="LoopIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 4V1L8 5l4 4V6c3.31 0 6 2.69 6 6 0 1.01-.25 1.97-.7 2.8l1.46 1.46C19.54 15.03 20 13.57 20 12c0-4.42-3.58-8-8-8zm0 14c-3.31 0-6-2.69-6-6 0-1.01.25-1.97.7-2.8L5.24 7.74C4.46 8.97 4 10.43 4 12c0 4.42 3.58 8 8 8v3l4-4-4-4v3z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="delete"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
          data-mui-internal-clone-element="true"
          tabindex="0"
          type="button"
          value="foo/bar/baz"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="DeleteIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
    >
      <div
        class="MuiListItemAvatar-root css-1e9lk82-MuiListItemAvatar-root"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-2s90m6-MuiAvatar-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="KeyIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M21 10h-8.35C11.83 7.67 9.61 6 7 6c-3.31 0-6 2.69-6 6s2.69 6 6 6c2.61 0 4.83-1.67 5.65-4H13l2 2 2-2 2 2 4-4.04L21 10zM7 15c-1.65 0-3-1.35-3-3s1.35-3 3-3 3 1.35 3 3-1.35 3-3 3z"
            />
          </svg>
        </div>
      </div>
      <div
        class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          foo/bar/baz
        </span>
        <p
          class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
        >
          <span
            class="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
          >
            sdlshdfkjlf
          </span>
          <br />
          <span
            class="MuiTypography-root MuiTypography-subtitle2 css-ati914-MuiTypography-root"
          >
            Last modified at: 
            10-10-2021 10:10
          </span>
        </p>
      </div>
      <button
        aria-label="copy"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-uof0vn-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element="true"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ContentCopyIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12V1zm3 4H8c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h11c1.1 0 2-.9 2-2V7c0-1.1-.9-2-2-2zm0 16H8V7h11v14z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-label="regenerate"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-uof0vn-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element="true"
        tabindex="0"
        type="button"
        value="foo/bar/baz"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="LoopIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 4V1L8 5l4 4V6c3.31 0 6 2.69 6 6 0 1.01-.25 1.97-.7 2.8l1.46 1.46C19.54 15.03 20 13.57 20 12c0-4.42-3.58-8-8-8zm0 14c-3.31 0-6-2.69-6-6 0-1.01.25-1.97.7-2.8L5.24 7.74C4.46 8.97 4 10.43 4 12c0 4.42 3.58 8 8 8v3l4-4-4-4v3z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-label="delete"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element="true"
        tabindex="0"
        type="button"
        value="foo/bar/baz"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="DeleteIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
